ls1intum/Artemis
# `Lectures`: Allow instructors to import existing lectures into courses

### Checklist
#### General


#### Server
#### Client

### Motivation and Context


Instructors usually hold Courses in a regular interval with little changes to the lecture content (https://github.com/ls1intum/Artemis/issues/4880).
Therefore, we introduce a process to import existing lectures from other Courses to the Course.

### Description

Using the import functionality, instructors can browse all Lectures they have access to.
In the modal, they can search and sort existing Lectures.
When clicking the import button, Artemis clones the Lecture including its Lecture Units and Attachments to the Course.
Exercise Units are excepted here, as there exists a separate process to import the different types of exercises on the Exercise Management page.
> This is one of my first pull requests on this project, so constructive feedback is very welcome.

### Steps for Testing

Prerequisites:
- 1 Instructor or Editor

1. Log in to Artemis
2. Navigate to Course Administration
3. Open a Course or create one where you are 
4. Create a Lecture, optionally with different Lecture Units and Attachments to the Lecture
5. Open or create another Course
6. Click the "Import Lecture" button on the Lecture Management page of that Course
7. Import a Lecture and verify Artemis also cloned the respective Lecture Units and Attachments

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| LectureImportService.java | 92% | 99% |
| LectureService.java | 85% | 95% |
| LectureResource.java | 79% | 97% |
| lecture-import.component.ts  | 76% | 92% |

### Screenshots



![screencast](https://user-images.githubusercontent.com/102237363/163780300-387865d7-9af6-43a8-bf9b-198e68bee3a2.gif)
